voltdbclient/conn.go

```diff
@@ -111,371 +88,89 @@ func (vc VoltConn) Close() (err error) {
+// current thread until that background thread has finished with all asynchronous requests.
+func (vc VoltConn) Drain() {
+	for _, nc := range vc.ncs {
+		nc.drain()
 	}
-	return newVoltConn(connInfo, tcpConn, tcpConn, *connData), nil
-}
-
-// Prepare creates a prepared statement for later queries or executions.
-// The Statement returned by Prepare is bound to this VoltConn.
-func (vc VoltConn) Prepare(query string) (driver.Stmt, error) {
-	stmt := newVoltStatement(&vc, query)
-	return *stmt, nil
 }
 
 // Exec executes a query that doesn't return rows, such as an INSERT or UPDATE.
 // Exec is available on both VoltConn and on VoltStatement.
 func (vc VoltConn) Exec(query string, args []driver.Value) (driver.Result, error) {
-	if !vc.isOpen() {
-		return nil, errors.New("Connection is closed")
-	}
-	handle := atomic.AddInt64(&qHandle, 1)
-	c := vc.nl().registerRequest(handle, false)
-	if err := vc.serializeQuery(vc.writer(), query, handle, args); err != nil {
-		vc.nl().removeRequest(handle)
-		return VoltResult{}, err
-	}
-	resp := <-c
-	rslt := resp.(VoltResult)
-	if err := rslt.getError(); err != nil {
-		return nil, err
-	}
-	return rslt, nil
+	nc := vc.nextNC()
+	return nc.exec(query, args)
 }
 
 // Exec executes a query that doesn't return rows, such as an INSERT or UPDATE.
 // ExecAsync is analogous to Exec but is run asynchronously.  That is, an
 // invocation of this method blocks only until a request is sent to the VoltDB
 // server.
 func (vc VoltConn) ExecAsync(resCons AsyncResponseConsumer, query string, args []driver.Value) error {
-	if !vc.isOpen() {
-		return errors.New("Connection is closed")
-	}
-	handle := atomic.AddInt64(&qHandle, 1)
-	c := vc.nl().registerRequest(handle, false)
-	vasr := newVoltAsyncResponse(vc, handle, c, false, resCons)
-	vc.registerAsync(handle, vasr)
-	if err := vc.serializeQuery(vc.writer(), query, handle, args); err != nil {
-		vc.nl().removeRequest(handle)
-		return err
-	}
-	return nil
+	nc := vc.nextNC()
+	return nc.execAsync(resCons, query, args)
+}
+
+// Prepare creates a prepared statement for later queries or executions.
+// The Statement returned by Prepare is bound to this VoltConn.
+func (vc VoltConn) Prepare(query string) (driver.Stmt, error) {
+	stmt := newVoltStatement(&vc, query)
+	return *stmt, nil
 }
 
 // Query executes a query that returns rows, typically a SELECT. The args are for any placeholder parameters in the query.
 func (vc VoltConn) Query(query string, args []driver.Value) (driver.Rows, error) {
-	if !vc.isOpen() {
-		return nil, errors.New("Connection is closed")
-	}
-	handle := atomic.AddInt64(&qHandle, 1)
-	c := vc.nl().registerRequest(handle, true)
-	if err := vc.serializeQuery(vc.writer(), query, handle, args); err != nil {
-		vc.nl().removeRequest(handle)
-		return VoltRows{}, err
-	}
-	resp := <-c
-	rows := resp.(VoltRows)
-	if err := rows.getError(); err != nil {
-		return nil, err
-	}
-	return rows, nil
+	nc := vc.nextNC()
```

Comment:
You should move this down to a diff class and keep client clean ala Distributor style otherwise this will get complicated fast with client affinity, RO queries from replica and any other optimizations we do.